assembly/uastc-generator.ts

```diff
@@ -0,0 +1,419 @@
+/**
+ * Generate UASTC data
+ * @param mode - UASTC Mode Index
+ * @param nBlocks - The total number of UASTC blocks
+ */
+export function generate(mode: i32, nBlocks: u32): i32 {
+  if (mode < 0 || mode > 18) return 0;
+
+  const totalBytes = nBlocks * 16;
+  if (<u32>(memory.size() - 1) * 65536 < totalBytes) return -1;
+
+  for (let seed: u64 = 0; seed < nBlocks; seed++) {
+    let r0: u64 = 0;
+
+    switch (mode) {
+      case 0:
+        r0 = 0x01; //0001
+
+        r0 |= crc64(0, seed) << (4 + 15);
+        break;
+      case 1:
+        r0 = 0x35; //110101
+
+        r0 |= crc64(0, seed) << (6 + 15);
+        break;
+      case 2:
+        r0 = 0x1D; //11101
+
+        r0 |= (seed % 30) << (5 + 15); // patterns
+        r0 |= crc64(0, seed) << (5 + 15 + 5);
+        break;
+      case 3:
+        r0 = 0x03; //00011
+
+        r0 |= (seed % 11) << (5 + 15); // patterns
+        r0 |= crc64(0, seed) << (5 + 15 + 4);
+        break;
+      case 4:
+        r0 = 0x13; //10011
+
+        r0 |= (seed % 30) << (5 + 15); // patterns
+        r0 |= crc64(0, seed) << (5 + 15 + 5);
+        break;
+      case 5:
+        r0 = 0x0B; //01011
+
+        r0 |= crc64(0, seed) << (5 + 15);
+        break;
+      case 6:
+        r0 = 0x1B; //11011
+
+        r0 |= (seed & 3) << (5 + 15); // ccs
+        r0 |= crc64(0, seed) << (5 + 15 + 2);
+        break;
+      case 7:
+        r0 = 0x07; //00111
+
+        r0 |= (seed % 19) << (5 + 15); // patterns
+        r0 |= crc64(0, seed) << (5 + 15 + 5);
+        break;
+      case 8:
+        r0 = 0x17; //10111
+
+        r0 |= crc64(0, seed) << 5;
+        break;
+      case 9:
+        r0 = 0x0F; //01111
+
+        r0 |= (seed % 30) << (5 + 23); // patterns
+        r0 |= crc64(0, seed) << (5 + 23 + 5);
+        break;
+      case 10:
+        r0 = 0x02; //010
+
+        r0 |= crc64(0, seed) << (3 + 17);
+        break;
+      case 11:
+        r0 = 0x00; //00
+
+        r0 |= (seed & 3) << (2 + 17); // ccs
+        r0 |= crc64(0, seed) << (2 + 17 + 2);
+        break;
+      case 12:
+        r0 = 0x06; //110
+
+        r0 |= crc64(0, seed) << (3 + 17);
+        break;
+      case 13:
+        r0 = 0x1F; //11111
+
+        r0 |= (seed & 3) << (5 + 23); // ccs
+        r0 |= crc64(0, seed) << (5 + 23 + 2);
+        break;
+      case 14:
+        r0 = 0x0D; //01101
+
+        r0 |= crc64(0, seed) << (5 + 23);
+        break;
+      case 15:
+        r0 = 0x05; //0000101
+
+        r0 |= crc64(0, seed) << (7 + 23);
+        break;
+      case 16:
+        r0 = 0x15; //010101
+
+        r0 |= (seed % 30) << (6 + 23); // patterns
+        r0 |= crc64(0, seed) << (6 + 23 + 5);
+        break;
+      case 17:
+        r0 = 0x25; //100101
+
+        r0 |= crc64(0, seed) << (6 + 23);
+        break;
+      case 18:
+        r0 = 0x09; //1001
+
+        r0 |= crc64(0, seed) << (4 + 15);
+        break;
+    }
+
+    const r1 = crc64(r0, seed);
+
+    const offset = 65536 + <u32>seed * 16;
```

Comment:
```suggestion
    const offset = 2048 + <u32>seed * 16;
```
`2048` = `65536 - 63488`

Reply:
Why? According to the docs, `--memoryBase` affects only memory segments (like static data or heap).

Reply:
You are right! Only for static data and managed heap)

Reply:
May be make sense do this like:
```ts
const offset = (ASC_MEMORY_BASE + 2048) + <u32>seed * 16;
```
WDYT?

Reply:
Not sure about that as `65536` is used explicitly in all other modules of this project as the user data offset. Maybe we should provide it with `asc -u USER_DATA_OFFSET=65536` everywhere. 

As we reserve the zeroth memory page for internal needs, `63488` was chosen to put generator's data to the very end of it so the generator and the transcoders could use the same memory without risking any overlap (although currently implemented UASTC modules use only up to 6kB of 64kB).